Support for multiple compression algorithms

dkorbelainen opened the pull request:
Support for multiple compression algorithms (LZ4, Zstandard, Brotli, besides the initial ZLIB & Dummy).

- Adds an enum for compressor types, stores the selected compressor type in the archive header, and ensures the correct compressor is automatically selected when opening an existing archive. 
- Serialization and deserialization of the compressor type are fully implemented and tested.

dkorbelainen committed:
feat: add LZ4, Zstandard, and Brotli compression support
- Add lz4, zstd, and brotli dependencies to vcpkg.json
- Implement LZ4, Zstandard, and Brotli compressors in compressor.c
- Add public API declarations for new compressors in compio.h
- Update CMakeLists.txt to link new compression libraries
- Add compressor_example demonstrating all compression algorithms

dkorbelainen committed:
feat: add compression type persistence and auto-detection
- Add compression_type field to archive header
- Implement auto-detection of compression algorithm on archive open
- Add helper functions for compressor type conversion
- Add compression persistence test example

dkorbelainen committed:
docs: add compression documentation

- Add API usage guide with compression examples
- Add compression examples README
- Add Doxygen documentation for LZ4, Zstandard, Brotli functions

dkorbelainen committed:
fix: address compression code review issues
- Add size validation for LZ4 to prevent data loss on int overflow
- Replace VLA with malloc/free in compressor example for safety
- Update error handling documentation for LZ4 functions

dkorbelainen commented on src/compressor.c:
```diff
@@ -77,4 +81,211 @@ void compio_build_zlib_compressor(compio_compressor* result) {
+ * @brief Compress data using LZ4 algorithm
+ *
+ * @param dst Destination buffer for compressed data
+ * @param dst_size Pointer to destination buffer size (input/output)
+ * @param src Source data buffer
+ * @param src_size Source data size in bytes
+ * @return 0 on success, -1 on error (sets errno to ENOBUFS if buffer too small)
+ */
+int lz4_compress(void* dst, uint64_t* dst_size, const void* src, uint64_t src_size) {
+    int compressed_size = LZ4_compress_default(
+        (const char*)src,
+        (char*)dst,
+        (int)src_size,
+        (int)*dst_size
+    );
```
Addressed.

mozhaa commented on src/compio.cpp:
```diff
@@ -56,8 +102,6 @@ compio_archive* compio_open_archive(const char* fp, const char* mode, const comp
     }
 
-    // if w+ passed as mode, we have to clear file contents (using w+)
-    //, otherwise we open with a+ mode to read and write
```
не убирай пж тут коммент, он тут объясняет как я выбираю режим для открытия

mozhaa commented on src/compio.cpp:
```diff
@@ -68,18 +112,36 @@ compio_archive* compio_open_archive(const char* fp, const char* mode, const comp
+    }
+
+    // Store the updated config
+    archive->config = new compio_config(config_copy);
```
мне кажется тут лучше не скрытно от пользователя менять тип сжатия, а вернуть код ошибки, если компрессор не совпадает

dkorbelainen replied:
Согласен

mozhaa commented on src/compio.cpp:
```diff
@@ -68,18 +112,36 @@ compio_archive* compio_open_archive(const char* fp, const char* mode, const comp
+    // Store the updated config
+    archive->config = new compio_config(config_copy);
 
-    // initialize allocator before btree, because btree uses allocator for creating root node
```
этот коммент тоже оставь, тут очень важно, что index после аллокатора создается (уже пару раз на этом баги были)

dkorbelainen committed:
fix: address code review feedback
- Add comment explaining archive open mode selection (w+ vs a+)
- Return error on compression type mismatch instead of silent modification
- Add comment about allocator initialization order before btree

mozhaa reviewed (approved):
Вроде норм
